File: R/FunMarkCor.R
#' Function to obtain the Functional Mark Correlation Function
#'
#' This function computes the Functional Mark Correlation Function as described in Comas et al (2009) and Comas et al. (2013)
#' We consider as a test function for the functional point pattern, L2 distances between functions
#'
#' @examples 
#'  using Data_Duke
#'  Functiona.mark.cor(
#'  
#' @param xy, object of class ppp (point pattern)
#' @param r, sequence of values to evaluate the Functional mark correlation function
#' @param rf, sequence of values for the function associated to each point
#' @param funpp is a matrix that contains de functions for each point, funct[,1] is the function for point 1
#' @param kernel, kernel function for both the pair correlation function and the Functional mark correlation function
#' @param bw bandwidth for the kernel function for the pair correlation function and Functional mark correlation function
#' @param lambda, point intensity
#' @param stoyan value for the epanechnikov kernel, default stoyan=0.15
#' @param correction for edge-effect for both the pair correlation function and the Functional mark correlation function
#' @return 
#' @author Carles Comas \email{carles.comas@udl.cat}
#' @useDynLib prova.fda, .registration=TRUE
#' @import spatstat
#' @export

FunMarkCor <- function(xy, r, rf, funpp, stoyan=0.15, kernel = "epanechnikov", bw, lambda, correction = "isotropic"){
        
         verifyclass(xy, "ppp")
         correc <- c("none", "isotropic")

         if (missing(funpp)) stop("the funpp matrix should be given")
         if (!inherits(funpp, "matrix")) stop("funct should be a matrix")

         if (missing(rf)) stop("the rf sequence should be given")
         if (!inherits(rf, "numeric")) stop("rf should be a numeric value")
         if (length(rf)<7) stop("rf should be a sequence of values larger than 6")

         indc <- match(correction, correc, nomatch = NA)
         if (any(cm <- is.na(indc))){
          nocm <- paste("unrecognised correction method:",paste(dQuote(correction[cm]),collapse=","))
          stop(nocm,call. = FALSE)
         }
         
         indc <- unique(indc)	
         edge <- rep(0, 2)
         edge[indc] <- 1	

         kern <- c("rectangular", "epanechnikov", "biweight")
         indk <- match(kernel, kern, nomatch = NA)
         if (any(km <- is.na(indk))){
         nokm <- paste("unrecognised kernel function:",paste(dQuote(kernel[km]),collapse=","))
         stop(nokm,call. = FALSE)
         }

         indk <- unique(indk)
         ker2 <- rep(0, 3)
         ker2[indk] <- 1

         win <- xy$window 
         n <- xy$n
         areaW <- area.owin(win)

         if(missing(lambda)) lambda <- n/areaW

         if (missing(bw) && (kernel == "epanechnikov") && length(lambda)== 1) {
          bw <- stoyan/sqrt(lambda)
         }
 
         if (missing(bw)){
            if((kernel != "epanechnikov") | lenght(lambda)>1) {
           bw <- bw.pcf(X = xy, kernel = kernel)[1] 
            }
         }
         if (!inherits(bw, "numeric")) stop("bw should be a numeric value")

        

         if (missing(r)){
         rect <- as.rectangle(win)
         maxd <- min(diff(rect$xrange), diff(rect$yrange))/4
         r <- seq(maxd*0.1, maxd, len = 301)[-1]
         r <- sort(r)
         }

         if(r[1] == 0) r <- r[-1]
         if (!inherits(r, "numeric")) stop("r should be a numeric value")
         if (length(r)<=1) stop("r should be a sequence of values")

         kernel <- c(kernel = kernel, bw = bw)
         
         n <- xy$n
         il <- seq(1, n)
         nr <- length(r)
         areaW <- area.owin(win)
         Ar<-areaW

       
         if (length(lambda) == 1) lambda <- rep(lambda, n)
    
         d <- pairdist(xy)
         if(correction == "isotropic") cor <- edge.Ripley(xy, d)
         if(correction == "none") cor <- 1

       
          nrf<-length(rf)
          minrf<-rf[1]

          na1<-c()
          for(i in 1:n){ 
            a1<-funpp[,i]
            na1[i]=length(a1[!is.na(a1)])
          }

          
          maxrf<-array(NA,c(n,n))
          dim(maxrf)=c(n,n)

          for(i in 1:n){
           for(j in 1:n){
            na<-min(na1[i],na1[j])
            maxrf[i,j]<-rf[na]
           }
          }


         FMCF<-rep(10,nr)
         storage.mode(FMCF) <- "double"
         
out<-.Fortran("FunMarkCorf1", funpp=as.double(funpp),d = as.double(d),
                       n = as.integer(n), r = as.double(r),
                       nr = as.integer(nr), nrf=as.integer(nrf), minrf=as.double(minrf),
                       maxrf=as.double(maxrf), ker2 = as.integer(ker2),
                       bw = as.double(bw), lambda = as.double(lambda),
                       cor = as.double(cor), edge = as.integer(edge),
                       Ar=as.double(Ar), out=(FMCF), PACKAGE = "prova.fda")
                      

invisible(return(list(FMCF = out$out, r = r, kernel = kernel, lambda = lambda)))
}

File: src/FunMarkCor.f
C
C     Author: Carles Comas
C     Date: 15th desember 2021 
C
C     Function  Functional.mark.cor.f to obtain the Functional Mark Correlation Function (FMCF)
C     
C

       SUBROUTINE FunMarkCorf1(funpp,d,n,r,nr,nrf,minrf,maxrf,
     &                 kr,bw,lamd,cor,edge,Ar,FMCF)


       IMPLICIT real*8 (a-h,o-z)

C      PARAMETERS
C      i, j, i1 internal counting variables, integer
C      TestF, matrix with Test function combining functions of the point pattern, test function: L2 distances
C      d, pairwise distance. double precision d(n,n)
C      n, number of points of the point pattern, interger
C      r, sequence of values, for the discret set of distances where the FMCF is evaluated, double precision
C      nr, the length of r, integer
C      nrf, the legnth of rf, integer
C      kr, value of the kernel function, integer value
C      bw, bandwidth value, real*8
C      lamd, point intensity, double precision
C      cor, value of the isotropic correction for point i wrt point j
C      edge, value for the edge-effect correcion, interger 
C      Ar, area of the window of observation, double precision
C      FMCF, Functional Mark Correlation Function, output

       INTEGER  i, j, i1, n, nr, kr, edge, nrf  
       DOUBLE PRECISION d(n,n), r(nr), lamd(n), cor(n,n)
       DOUBLE PRECISION TestF(n,n), FMCF(nr), PCF(nr)
       DOUBLE PRECISION fxs(nrf), funpp(nrf,n),maxrf(n,n)
       REAL*8 Ar, bw, dij, kerns, ExpF, a, b, minrf
       DIMENSION kr(3), edge(2)
     
       
     

C      PAIR CORRELATION FUNCTION

       DO i1=1, nr
        PCF(i1)=0.0
         DO i=1,n
          DO j=1,n
           IF (j.ne.i) THEN
            dij=d(i,j)
              IF (kr(1).eq.1) THEN
c             kerns=1
               kerns=boxkernel((r(i1)-dij)/bw,bw)
              ELSE IF (kr(2).eq.1) THEN
               kerns=ekernel((r(i1)-dij)/bw,bw)
              ELSE IF (kr(3).eq.1) THEN
              kerns=qkernel((r(i1)-dij)/bw,bw)
              END IF

              IF (kerns.ne.0) THEN
C     none   
                IF (edge(1).eq.1) THEN
                 wij=kerns/(lamd(i)*lamd(j)*r(i1))
                 PCF(i1)=PCF(i1)+wij 
                END IF                          
C    isotropic
                IF (edge(2).eq.1) THEN                  
                 wij=(kerns*cor(i,j))/(lamd(i)*lamd(j)*r(i1))
                 PCF(i1)=PCF(i1)+wij
                END IF

              END IF
           END IF
         END DO
        END DO
        PCF(i1)=PCF(i1)/(2*3.141592654*Ar)
       END DO

C   TEST FUNCTION

        DO i=1, n
         DO j=1, n
           IF (j.ne.i) THEN 
             DO i1=1, nrf
             fxs(i1)=(funpp(i1,i)-funpp(i1,j))**2
             ENDDO
             TestF(i,j)=sqrt(SIMP(fxs, minrf, maxrf(i,j),nrf))
           ENDIF
         ENDDO
        ENDDO


C   FUNCTIONAL MARK CORRELATION FUNCTION

      ExpF=ETestF(n,TestF)

      DO i1=1, nr
        FMCF(i1)=0.0
         DO i=1,n
          DO j=1,n
           IF (j.ne.i) THEN
            dij=d(i,j)
              IF (kr(1).eq.1) THEN
               kerns=boxkernel((r(i1)-dij)/bw,bw)
              ELSE IF (kr(2).eq.1) THEN
               kerns=ekernel((r(i1)-dij)/bw,bw)
              ELSE IF (kr(3).eq.1) THEN
              kerns=qkernel((r(i1)-dij)/bw,bw)
              END IF

              IF (kerns.ne.0) THEN
C     none   
                IF (edge(1).eq.1) THEN
             wij1=(kerns*TestF(i,j))
             wij2=lamd(i)*lamd(j)*r(i1)*ExpF*PCF(i1)
             wij=wij1/wij2     
                 FMCF(i1)=FMCF(i1)+wij 
                END IF                          
C    isotropic
                IF (edge(2).eq.1) THEN                  
            wij1=kerns*TestF(i,j)*cor(i,j)
            wij2=lamd(i)*lamd(j)*r(i1)*ExpF*PCF(i1)
            wij=wij1/wij2
                 FMCF(i1)=FMCF(i1)+wij
                END IF

              END IF
           END IF
         END DO
        END DO
        FMCF(i1)=FMCF(i1)/(2*3.141592654*Ar)
       END DO

        RETURN
        END
       


       
CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCc
C
C     functions called by :
C     -----------------------------------------
C
C     * boxkernel, ekernel, qkernel
C
CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC

C--------------------------------------------------------------------
C
C     boxkernel
C
C--------------------------------------------------------------------

       FUNCTION boxkernel(x,h)

       IMPLICIT REAL*8 (a-h,o-z)

       REAL*8 x, h

       IF (abs(x).le.1d0) THEN
         boxkernel=1d0/2d0
       ELSE
         boxkernel=0d0
       END IF
       boxkernel=boxkernel/h
       RETURN
       END

C--------------------------------------------------------------------
C
C     Epanechnikov kernel
C
C--------------------------------------------------------------------

       FUNCTION ekernel(x,h)

     
       IMPLICIT REAL*8 (a-h,o-z)
       REAL*8 x, h

       IF (abs(x).le.1) THEN
           ekernel=(3d0/(4d0*h))*(1-x**2)
       ELSE
          ekernel=0.0
       ENDIF

C       IF (abs(x).le.1d0) THEN
C           ekernel=(3d0/4d0)*(1-x**2)
C       ELSE
C           ekernel=0d0
C       END IF
C       ekernel=ekernel/h

        RETURN
        END

 

C--------------------------------------------------------------------
C
C     quartic (biweight) kernel
C
C--------------------------------------------------------------------

       FUNCTION qkernel(x,h)

       IMPLICIT REAL*8 (a-h,o-z)
       REAL*8 x, h

       IF (abs(x).le.1d0) THEN
           qkernel=(15d0/16d0)*(1-x**2)**2
       ELSE
           qkernel=0d0
       END IF
       qkernel=qkernel/h

       RETURN
       END

C--------------------------------------------------------------------
C
C     Expected Test Function
C
C--------------------------------------------------------------------

      FUNCTION ETestF(n,TestF)

      IMPLICIT REAL*8 (a-h,o-z)
      INTEGER n
      DOUBLE PRECISION TestF(n,n)

       ETestF=0.0
       DO i=1,n
          DO j=1,n
           IF (j.ne.i) THEN
           ETestF=ETestF+TestF(i,j)
           ENDIF
         ENDDO
       ENDDO
       ETestF=ETestF/(n**2-n)
       RETURN
       END

C--------------------------------------------------------------------
C
C     Integral Function using Simpson
C
C--------------------------------------------------------------------

        FUNCTION SIMP(fxs,a,b,nrf)
        IMPLICIT REAL*8 (a-h,o-z)
        INTEGER nrf,c1,c2
        DOUBLE PRECISION fxs(nrf)
        REAL*8 a, b, h
        
         h=(b-a)/(nrf-1.0)	
         SIMP=3.0*(fxs(1)+fxs(nrf))/8.0+7.0*(fxs(2)+
     &    fxs(nrf-1))/6.0+23.0*(fxs(3)+fxs(nrf-2))/24.0
         c1=4
         c2=nrf-3
         DO i=c1, c2
          SIMP=SIMP+fxs(i)
         ENDDO
         SIMP=SIMP*h
         RETURN
         END




